### PR TITLE
Ensure that child items are deleted when their parents are

### LINKED ISSUE #5: Files are orphaned when deleting their parent folder
If you delete a folder and recreate it, it doesn't remove any of the children of that folder. 
image
```
If IModifiableFolder.CreateFolderAsync is called but the name is already in use, then when overwrite is true the created folder should be empty.
```

The code for this can be found [here](https://github.com/Arlodotexe/OwlCore.Storage.SharpCompress/blob/706f7eb030d7a25f0b03cb5127781e33faa990d1/src/ArchiveFolder.cs#L44).

_Originally posted by @Arlodotexe in https://github.com/yoshiask/OwlCore.Storage.SharpCompress/issues/4#issuecomment-2004779589_

### PR DESCRIPTION
**Pull request by yoshiask**
Fixes #5

### CODE REVIEW

File: src/ArchiveFolder.cs [hunk 1]
```diff
@@ -74,25 +83,19 @@ public async Task<IChildFile> CreateFileAsync(string name, bool overwrite = fals
 
     protected override ReadOnlyArchiveFolder WrapSubfolder(string name) => new ArchiveFolder(this, name);
 
-    protected async Task RemoveSubfolder(IArchiveEntry entry, CancellationToken cancellationToken = default)
```
Comment (yoshiask) [inline src/ArchiveFolder.cs:77]: This method wasn't used anywhere, so it didn't seem worth updating it to fix this bug.

File: src/ArchiveFolder.cs [hunk 2]
```diff
@@ -33,15 +33,24 @@ public async Task<IChildFolder> CreateFolderAsync(string name, bool overwrite =
...
+            return folder;
+            
+        // Folder already exists and caller wants to overwrite it,
+        // so get the parent and attempt to delete the existing
+        // one before creating a new one.
+        var parent = await folder.GetParentAsync(cancellationToken);
+        if (parent is not IModifiableFolder modifiableParent)
+            throw new IOException($"A folder with the name '{name}' already exists in parent folder " +
+                                  $"'{parent?.Id ?? null}' and is not modifiable.");
+        await modifiableParent.DeleteAsync(folder, cancellationToken);
```
Comment (yoshiask) [inline src/ArchiveFolder.cs:53]: This is the real change in this method: actually call `DeleteAsync` to get the full recursive removal of the folder *before* creating it again.

File: src/ArchiveFolder.cs [hunk 3]
```diff
@@ -74,25 +83,19 @@ public async Task<IChildFile> CreateFileAsync(string name, bool overwrite = fals
...
     {
         cancellationToken.ThrowIfCancellationRequested();
-
+        
+        // Remove this and all child entries from archive.
+        // Force enumeration with .ToList() since we're modifying the collection.
         var archive = await OpenWritableArchiveAsync(cancellationToken);
+        var entries = archive.Entries.ToList();
+        foreach (var entry in entries)
+            if (entry.Key == key || IsChild(entry.Key, key))
```
Comment (yoshiask) [inline src/ArchiveFolder.cs:95]: The key change in here is that we remove *every* child entry, not just the requested folder.

**Review by Arlodotexe — APPROVED** [review]
Looks like tests are passing now!

**Pull request merged**